src/DocNet/NavigationLevel.cs

```diff
@@ -31,198 +31,325 @@
+                                                    this.Name));
+                    }
+                }
+                // then the elements in the container. Index elements are skipped here.
+                foreach (var element in this.Value)
+                {
+                    fragments.Add(element.GenerateToCFragment(navigatedPath, relativePathToRoot));
+                }
+                fragments.Add("</ul>");
+            }
+            else
+            {
+                // just a link
+                fragments.Add(string.Format("<span class=\"navigationgroup\"><i class=\"fa fa-caret-right\"></i> <a href=\"{0}{1}\">{2}</a></span>",
+                                            relativePathToRoot, HttpUtility.UrlPathEncode(this.TargetURL), this.Name));
+            }
+            if (!this.IsRoot)
+            {
+                fragments.Add("</li>");
+            }
+            return string.Join(Environment.NewLine, fragments.ToArray());
+        }
+
+        private NavigationLevel CreateGeneratedLevel(string path)
+        {
+            var root = new NavigationLevel(_rootDirectory)
+            {
+                ParentContainer = this
+            };
+
+            foreach (var mdFile in Directory.GetFiles(path, "*.md", SearchOption.TopDirectoryOnly))
+            {
+                var name = FindTitleInMdFile(mdFile);
+                if (string.IsNullOrWhiteSpace(name))
+                {
+                    continue;
+                }
+
+                var relativeFilePath = GetRelativePath(_rootDirectory, mdFile);
+
+                var item = new SimpleNavigationElement
+                {
+                    Name = name,
+                    Value = relativeFilePath,
+                    ParentContainer = root
+                };
+
+                root.Value.Add(item);
+            }
+
+            foreach (var directory in Directory.GetDirectories(path, "*", SearchOption.TopDirectoryOnly))
+            {
+                var directoryInfo = new DirectoryInfo(directory);
+
+                var subDirectoryNavigationElement = CreateGeneratedLevel(directory);
+
+                subDirectoryNavigationElement.Name = directoryInfo.Name;
+                subDirectoryNavigationElement.ParentContainer = root;
+
+                root.Value.Add(subDirectoryNavigationElement);
+            }
+
+            return root;
+        }
+
+        private string GetRelativePath(string origin, string fullPath)
+        {
+            var pathUri = new Uri(fullPath);
+
+            if (!origin.EndsWith(Path.DirectorySeparatorChar.ToString()))
+            {
+                origin += Path.DirectorySeparatorChar;
+            }
+
+            var folderUri = new Uri(origin);
+            return Uri.UnescapeDataString(folderUri.MakeRelativeUri(pathUri).ToString().Replace('/', Path.DirectorySeparatorChar));
+        }
+
+        private string FindTitleInMdFile(string path)
+        {
```

Comment:
I am not extremely happy with this one either, but looked like the least-worst solution.